arxivdigest/core/mail/digest_mail.py

```diff
@@ -12,46 +13,50 @@
-from arxivdigest.core.config import config_email
-from arxivdigest.core.config import config_interleave
-from arxivdigest.core.config import config_web_address
+from arxivdigest.core.config import (
+    config_email,
+    config_interleave,
+    config_web_address,
+)
 from arxivdigest.core.database import users_db
 from arxivdigest.core.mail.mail_server import MailServer
 
-RECOMMENDATIONS_PER_USER = config_interleave.get('recommendations_per_user')
-SYSTEMS_PER_USER = config_interleave.get('systems_multileaved_per_user')
-BATCH_SIZE = config_interleave.get('users_per_batch')
+RECOMMENDATIONS_PER_USER = config_interleave.get("recommendations_per_user")
+SYSTEMS_PER_USER = config_interleave.get("systems_multileaved_per_user")
+BATCH_SIZE = config_interleave.get("users_per_batch")
 BASE_URL = config_web_address
-ARTICLES_PER_DATE_IN_MAIL = config_interleave.get('articles_per_date_in_email')
+ARTICLES_PER_DATE_IN_MAIL = config_interleave.get("articles_per_date_in_email")
 
 
 def send_digest_mail():
     """Sends emails to users about new recommendations."""
     article_data = get_article_data()
-    server = MailServer(**config_email)
     n_users = users_db.get_number_of_users()
     n_batches = math.ceil(n_users / BATCH_SIZE)
     for offset in range(0, n_users, BATCH_SIZE):
-        batch = int(offset / BATCH_SIZE) + 1
-        logging.info('Sending mails for batch {} of {}.'.format(batch,
-                                                                n_batches))
-
-        mail_batch, trace_batch = create_mail_batch(offset, article_data)
-        if not mail_batch:
-            logging.info('Batch {} was empty.'.format(batch))
-            continue
+        with closing(MailServer(**config_email)) as server:
+            batch = int(offset / BATCH_SIZE) + 1
+            logging.info(
+                "Sending mails for batch {} of {}.".format(batch, n_batches)
+            )
 
-        # mark all articles as seen in database and send all mails
-        insert_mail_trackers(trace_batch)
-        for mail in mail_batch:
-            server.send_mail(**mail)
-        logging.info('Batch {} sent.'.format(batch))
+            mail_batch, trace_batch = create_mail_batch(offset, article_data)
+            if not mail_batch:
+                logging.info("Batch {} was empty.".format(batch))
+                continue
 
-    server.close()
+            # mark all articles as seen in database and send all mails
+            insert_mail_trackers(trace_batch)
+            for mail in mail_batch:
+                try:
+                    server.send_mail(**mail)
+                except smtplib.SMTPSenderRefused:
+                    logging.error(f"Failed sending email")
+            logging.info("Batch {} sent.".format(batch))
 
 
 def create_mail_batch(offset, article_data):
-    """ Creates a batch of emails for users starting from 'start_id'.
+    """Creates a batch of emails for users starting from 'start_id'.
```

Comment:
Nit: "...starting for a given ID." (there is no start_id)

Reply:
Done